### DISCUSSION TITLE
Making a mechanical (gears etc) object. Which settings are the most effective for improving the accuracy?

### DISCUSSION THREAD

**Discussion by DustStock** [jrouwe/JoltPhysics#1199] [Q&A] [answered]
I am using Jolt in Godot, but I am looking for some general tips regarding the physics simulation. I have an object which has several mechanical interactions such as gears and other parts that push and pull on each other - similar to a lock or clock.

In general, which settings are the most important to adjust to increase the accuracy of these interactions?

I have some instances where the objects will clip or push through each other. What are the best settings to adjust to prevent this?

**Comment by jrouwe** [comment, marked as answer]
The most accurate simulation will be to use Jolt's [GearConstraint](https://www.youtube.com/watch?v=3w5SgElroBw) or [RackAndPinionConstraint](https://www.youtube.com/watch?v=e588KG-ZSxc). These are however not exposed in godot-jolt by default, so you would have to do some C++ programming to expose them in a similar way as e.g. the custom [JoltHingeJoint3D](https://github.com/godot-jolt/godot-jolt/blob/master/src/joints/jolt_hinge_joint_3d.cpp) class. Note that there is a difference: GearConstraint applies to 2 HingeConstraints and not to 2 Bodies like JoltHingeJoint3D, so it is not a straight copy paste.

If you can't or don't want to modify godot-jolt then I'd say it is quite difficult to accurately model gears. I'd probably look at simplifying the physics geometry compared to the visual geometry by reducing the number of teeth. If you e.g. make the teeth on both sides 2x as big then their ratio remains the same and this would probably help reduce clipping. Other than that you can increase simulation frequency and go to the Jolt 3D Settings to reduce "Contact Speculative Distance" and "Contact Allowed Penetration".

**Reply by DustStock** [reply]
Starting from default settings, I was able to get great behavior from the sim following your advice and 

- increasing simulation frequency 
- reducing "Contact Speculative Distance" and 
- reducing  "Contact Allowed Penetration".

It works great now, thanks!

**Comment by DustStock** [comment]
My apologies for cross posting, I didn't realize it was the same audience and wasn't sure which one would be better. 

And thanks for the reply, I have spent a lot of time tweaking settings to get better results - but I really need to step back an re-evaluate / optimize them since I am still having some issues.

What I am going for is a physics based sim where the user can make their own trinkets and mechanisms. So, I have avoided adding contraints and joints as I would like the behavior to be more emergent.

And, actually, the gears in my test scene have actually worked really well. I haven't had trouble with them...yet - only one test scene with them.

The parts I have the most trouble with are the parts that push and/or slide (especially when the mechanism is seized). These parts will jostle can clip through one another.

I am going to reset all of my settings tweaks and restart by lowering both the contact speculative distance and allowed penetration.

When you say "simulation frequency" this is Physics Ticks per Second, correct?

I have also experimented changing these:

Max Physics Steps per Frame
CCD / Movement Threshold / Max penetration
Kinematics Recovery Iterations / Amount
Solver Velocity and Position Iterations, Position Correction
Use Shape Margins
USe Enhanced Internal Edge Removal

Do you have any thoughts or tips regarding those? Are any of these things helpful in my situation? Are any of these redundant if I am cranking up the simulation frequency anyway?

**Reply by jrouwe** [reply]
> When you say "simulation frequency" this is Physics Ticks per Second, correct?

Correct

> Max Physics Steps per Frame

This one limits the amount of physics steps if the FPS is too low to keep up. If you have a good FPS this should not matter.

> CCD / Movement Threshold / Max penetration

This should only do something for fast moving objects. It sounds like your objects are not moving fast, so I don't expect any benefits.

> Kinematics Recovery Iterations / Amount

These are values are used only by a specific custom collision test and not by the simulation. I would expect this to not matter.

> Solver Velocity and Position Iterations, Position Correction

These settings do affect the quality of the simulation, especially a higher number of solver velocity steps increases quality of the constraints.

> Use Shape Margins

This increases the collision precision at a performance cost. I would not expect this to make a big difference, but you can try.

> USe Enhanced Internal Edge Removal

This setting specifically improves objects sliding over a polygon terrain. From what you've told me so far this is not your case.

Perhaps you can post a video of what your scene looks like?

**Reply by DustStock** [reply]
Thanks for the detailed information.

I am going to revisit my settings. If I continue to have undesirable effects, I will capture a video to share of a simple scene.
